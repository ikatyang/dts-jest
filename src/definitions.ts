import * as _ts from 'typescript';

// tslint:disable-next-line:no-var-requires
export const package_name = require('../package.json').name;

export const config_namespace = '_dts_jest_';
export const runtime_namespace = '_dts_jest_runtime_';

export const env_root_dir = 'DTS_JEST_ROOT_DIR';

export const runtime_indent_spaces = 2;

export const trigger_regex = /^\s*\/\/\s*@dts-jest\b(:?\S*)\s*(.+)?\s*$/;
export enum TriggerMatchIndex {
  Input,
  Flags,
  Description,
}
export type TriggerMatchArray = [string, string, string | undefined];

export enum AssertionFlag {
  Snapshot = ':snapshot',
  Show = ':show',
  Pass = ':pass',
  Fail = ':fail',
}
export enum ActualAssertionFlag {
  Error = ':error',
  NoError = ':no-error',
  Question = '?',
}

export enum TestFlag {
  Test = ':test',
  Only = ':only',
  Skip = ':skip',
}
export enum TestMethod {
  Test = 'test',
  Only = 'test.only',
  Skip = 'test.skip',
}
export const group_flag = ':group';
export enum GroupMethod {
  Test = 'describe',
  Only = 'describe.only',
  Skip = 'describe.skip',
}

export interface Target {
  /**
   * 0-based
   */
  line: number;
  expression: string;
  description?: string;
  group?: Group;
}

export interface Snapshot {
  inference?: string;
  diagnostic?: string;
}
export interface Expected {
  value: string;
}

export interface TriggerInfo {
  flags: AssertionFlag[];
  method: TestMethod;
}
export interface Trigger extends Target, TriggerInfo {
  start: number;
  end: number;
}
export interface ActualTrigger extends Trigger, Expected {}

export interface GroupInfo {
  method: GroupMethod;
}
export interface Group extends GroupInfo {
  title: string;
}

export type TriggerOrGroupInfo =
  | ({ is_group: false } & TriggerInfo)
  | ({ is_group: true } & GroupInfo);

export interface Result extends Target, Snapshot {}
export interface ActualResult extends Target, Expected {}

export interface JestConfig {
  rootDir: string;
  globals: { [K in typeof config_namespace]?: RawConfig };
  _dts_jest_internal_test_?: boolean;
}

export interface RawConfig {
  tsconfig?: string | Record<string, _ts.CompilerOptionsValue>;
  type_format?: _ts.TypeFormatFlags;
  typescript?: string;
}

export interface FormattedConfig {
  tsconfig: _ts.CompilerOptions;
  type_format: _ts.TypeFormatFlags;
  typescript: typeof _ts;
  typescript_path: string;
}

export type JestSnapshot = Record<string, string>;
