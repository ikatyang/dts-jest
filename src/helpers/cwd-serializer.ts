export const cwd_serializer: jest.SnapshotSerializerPlugin = {
  print: (value: string, serializer) =>
    serializer(value.replace(new RegExp(process.cwd(), 'g'), '<cwd>')),
  test: (value: any) =>
    typeof value === 'string' && value.indexOf(process.cwd()) !== -1,
};
