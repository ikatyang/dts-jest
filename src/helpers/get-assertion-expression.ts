import {
  runtime_namespace,
  IRuntimeReportParameters,
  IRuntimeSnapshotParameters,
  TriggerKind,
} from '../definitions';

export const get_assertion_expression = (kind: TriggerKind, line: number) => {
  if (kind === TriggerKind.Show) {
    const report_parameters: IRuntimeReportParameters = {line};
    const report_value = `${runtime_namespace}.report(${JSON.stringify(report_parameters)})`;
    return `console.log(${report_value})`;
  }
  const snapshot_parameters: IRuntimeSnapshotParameters = {line};
  const snapshot_value = `${runtime_namespace}.snapshot(${JSON.stringify(snapshot_parameters)})`;
  return `expect(${snapshot_value}).toMatchSnapshot()`;
};
