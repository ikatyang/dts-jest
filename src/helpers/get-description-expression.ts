import {
  runtime_namespace,
  IRuntimeDescriptionParameters,
} from '../definitions';

export const get_description_expression = (line: number) => {
  const parameters: IRuntimeDescriptionParameters = {
    line,
  };
  return `${runtime_namespace}.description(${JSON.stringify(parameters)})`;
};
