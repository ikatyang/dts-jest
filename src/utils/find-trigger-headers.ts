import * as _ts from 'typescript';
import {
  trigger_header_regex,
  GroupMethod,
  TestMethod,
  TriggerGroup,
  TriggerHeader,
  TriggerHeaderFlags,
  TriggerHeaderMatchIndex,
} from '../definitions';
import { for_each_comment } from './for-each-comment';
import { get_comment_content } from './get-comment-content';
import { parse_trigger_header_flags } from './parse-trigger-header-flags';

export const find_trigger_headers = (
  source_file: _ts.SourceFile,
  ts: typeof _ts,
) => {
  const headers: TriggerHeader[] = [];

  let last_group: TriggerGroup | undefined;

  for_each_comment(
    source_file,
    (comment, scanner) => {
      const match = get_comment_content(comment).match(trigger_header_regex);

      if (match === null) {
        return;
      }

      const description = match[TriggerHeaderMatchIndex.Description];
      const flags = parse_trigger_header_flags(
        match[TriggerHeaderMatchIndex.Flags],
      );

      const start = scanner.getTokenPos();
      const { line } = source_file.getLineAndCharacterOfPosition(start);

      if (flags & TriggerHeaderFlags[':group']) {
        last_group = {
          line,
          method: get_group_method(flags),
          description: description.length === 0 ? undefined : description,
        };
      } else {
        headers.push({
          line,
          flags,
          method: get_test_method(flags),
          description: description.length === 0 ? undefined : description,
          group: last_group,
        });
      }
    },
    ts,
  );

  return headers;
};

function get_test_method(flag: TriggerHeaderFlags) {
  return flag & TriggerHeaderFlags[':only']
    ? TestMethod.Only
    : flag & TriggerHeaderFlags[':skip'] ? TestMethod.Skip : TestMethod.Test;
}

function get_group_method(flag: TriggerHeaderFlags) {
  return flag & TriggerHeaderFlags[':only']
    ? GroupMethod.Only
    : flag & TriggerHeaderFlags[':skip'] ? GroupMethod.Skip : GroupMethod.Test;
}
