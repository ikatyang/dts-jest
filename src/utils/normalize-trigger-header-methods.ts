import {
  GroupMethod,
  TestMethod,
  TriggerGroup,
  TriggerHeader,
} from '../definitions';

export const normalize_trigger_header_methods = (headers: TriggerHeader[]) => {
  const has_only_method = headers.some(
    header =>
      header.method === TestMethod.Only ||
      (header.group !== undefined && header.group.method === GroupMethod.Only),
  );

  if (!has_only_method) {
    return;
  }

  headers.forEach(header => {
    if (
      header.method === TestMethod.Test &&
      (header.group === undefined || header.group.method === GroupMethod.Test)
    ) {
      header.method = TestMethod.Skip;
    }
  });

  const group_map = new Map<undefined | TriggerGroup, TriggerHeader[]>();
  headers.forEach(header => {
    set_array_map(group_map, header.group, header);
  });
  group_map.forEach((grouped_triggers, group) => {
    if (group !== undefined) {
      normalize_grouped_trigger_methods(grouped_triggers);
    }
  });

  headers.forEach(header => {
    if (header.method === TestMethod.Only) {
      header.method = TestMethod.Test;
    }
  });
};

function set_array_map<T, U>(map: Map<T, U[]>, key: T, value: U) {
  const values = map.get(key);
  if (values === undefined) {
    map.set(key, [value]);
  } else {
    values.push(value);
  }
}

function normalize_grouped_trigger_methods(headers: TriggerHeader[]) {
  const test_headers: TriggerHeader[] = [];
  const only_headers: TriggerHeader[] = [];

  const is_skip_group = headers[0].group!.method === GroupMethod.Skip;

  headers.forEach(header => {
    if (header.method === TestMethod.Test) {
      test_headers.push(header);
    } else if (header.method === TestMethod.Only) {
      only_headers.push(header);
    }
  });

  if (is_skip_group || only_headers.length !== 0) {
    test_headers.forEach(header => (header.method = TestMethod.Skip));
  }
}
