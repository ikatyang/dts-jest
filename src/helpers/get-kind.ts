import {TriggerKind} from '../definitions';

export const get_kind = (flag?: string) => {
  switch (flag) {
    case ':show':
      return TriggerKind.Show;
    case ':skip':
      return TriggerKind.Skip;
    case ':only':
      return TriggerKind.Only;
    case undefined:
      return TriggerKind.None;
    default:
      throw new Error(`Unexpected flag ${JSON.stringify(flag)}`);
  }
};
