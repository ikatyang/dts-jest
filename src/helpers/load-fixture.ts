import * as fs from 'fs';
import * as path from 'path';
import * as _ts from 'typescript';
import { create_source_file } from '../utils/create-source-file';

export const get_fixture_filename = (id: string) =>
  path.resolve(__dirname, '../../fixtures', id);

export const load_fixture = (id: string) =>
  fs.readFileSync(get_fixture_filename(id), 'utf8');

export const load_fixture_source_file = (id: string, ts: typeof _ts) => {
  const source = load_fixture(id);
  return create_source_file(id, source, ts);
};
