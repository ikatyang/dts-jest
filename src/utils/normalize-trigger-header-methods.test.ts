import * as ts from 'typescript';
import { TestMethod, TriggerHeader } from '../definitions';
import { load_fixture_source_file } from '../helpers/load-fixture';
import { find_trigger_headers } from './find-trigger-headers';
import { normalize_trigger_header_methods } from './normalize-trigger-header-methods';

it('should not be affected without only', () => {
  const headers = get_normalized_headers('no-only');
  expect(get_headers_info(headers)).toMatchSnapshot();
  expect(get_header_only_count(headers)).toBe(0);
});

it('should normalize correctly with group-only', () => {
  const headers = get_normalized_headers('group-only');
  expect(get_headers_info(headers)).toMatchSnapshot();
  expect(get_header_only_count(headers)).toBe(0);
});

it('should normalize correctly with trigger-only', () => {
  const headers = get_normalized_headers('trigger-only');
  expect(get_headers_info(headers)).toMatchSnapshot();
  expect(get_header_only_count(headers)).toBe(0);
});

it('should normalize correctly with trigger-only in group-only', () => {
  const headers = get_normalized_headers('trigger-only-in-group-only');
  expect(get_headers_info(headers)).toMatchSnapshot();
  expect(get_header_only_count(headers)).toBe(0);
});

it('should normalize correctly with trigger-only in group-skip', () => {
  const headers = get_normalized_headers('trigger-only-in-group-skip');
  expect(get_headers_info(headers)).toMatchSnapshot();
  expect(get_header_only_count(headers)).toBe(0);
});

it('should normalize correctly with multi group-only', () => {
  const headers = get_normalized_headers('multi-group-only');
  expect(get_headers_info(headers)).toMatchSnapshot();
  expect(get_header_only_count(headers)).toBe(0);
});

it('should normalize correctly with multi trigger-only', () => {
  const headers = get_normalized_headers('multi-trigger-only');
  expect(get_headers_info(headers)).toMatchSnapshot();
  expect(get_header_only_count(headers)).toBe(0);
});

it('should normalize correctly with multi group-only and trigger-only', () => {
  const headers = get_normalized_headers('multi-mixed-only');
  expect(get_headers_info(headers)).toMatchSnapshot();
  expect(get_header_only_count(headers)).toBe(0);
});

function get_normalized_headers(id: string) {
  const source_file = load_fixture_source_file(
    `normalize-trigger-header-methods/${id}.ts`,
    ts,
  );
  const headers = find_trigger_headers(source_file, ts);
  normalize_trigger_header_methods(headers);
  return headers;
}

function get_header_only_count(headers: TriggerHeader[]) {
  return headers.filter(header => header.method === TestMethod.Only).length;
}

function get_headers_info(headers: TriggerHeader[]) {
  return headers.map(header => `${header.description} -> ${header.method}`);
}
