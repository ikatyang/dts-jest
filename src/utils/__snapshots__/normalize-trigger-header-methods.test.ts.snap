// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should normalize correctly with group-only 1`] = `
Array [
  "1 -> test.skip",
  "2 -> test.skip",
  "A1 -> test",
  "A2 -> test",
  "B1 -> test.skip",
  "B2 -> test.skip",
]
`;

exports[`should normalize correctly with multi group-only 1`] = `
Array [
  "1 -> test.skip",
  "2 -> test.skip",
  "A1 -> test",
  "A2 -> test",
  "B1 -> test.skip",
  "B2 -> test.skip",
  "C1 -> test",
  "C2 -> test",
]
`;

exports[`should normalize correctly with multi group-only and trigger-only 1`] = `
Array [
  "1 -> test.skip",
  "2 -> test",
  "3 -> test",
  "A1 -> test.skip",
  "A2 -> test.skip",
  "B1 -> test.skip",
  "B2 -> test.skip",
  "C1 -> test",
  "C2 -> test",
]
`;

exports[`should normalize correctly with multi trigger-only 1`] = `
Array [
  "1 -> test.skip",
  "2 -> test",
  "3 -> test",
  "A1 -> test.skip",
  "A2 -> test.skip",
  "B1 -> test.skip",
  "B2 -> test.skip",
]
`;

exports[`should normalize correctly with trigger-only 1`] = `
Array [
  "1 -> test.skip",
  "2 -> test",
  "A1 -> test.skip",
  "A2 -> test.skip",
  "B1 -> test.skip",
  "B2 -> test.skip",
]
`;

exports[`should normalize correctly with trigger-only in group-only 1`] = `
Array [
  "1 -> test.skip",
  "2 -> test.skip",
  "A1 -> test",
  "A2 -> test.skip",
  "B1 -> test.skip",
  "B2 -> test.skip",
]
`;

exports[`should normalize correctly with trigger-only in group-skip 1`] = `
Array [
  "1 -> test.skip",
  "2 -> test.skip",
  "A1 -> test",
  "A2 -> test.skip",
  "B1 -> test.skip",
  "B2 -> test.skip",
]
`;

exports[`should not be affected without only 1`] = `
Array [
  "1 -> test",
  "2 -> test",
  "A1 -> test",
  "A2 -> test",
  "B1 -> test",
  "B2 -> test",
]
`;
